base/special/exp.jl

```diff
@@ -1,65 +1,198 @@
-# Based on FDLIBM http://www.netlib.org/fdlibm/e_exp.c
-# which is made available under the following licence
-
-## Copyright (C) 2004 by Sun Microsystems, Inc. All rights reserved. Permission
-## to use, copy, modify, and distribute this software is freely granted,
-## provided that this notice is preserved.
-
-# Method
-# 1. Argument reduction: Reduce x to an r so that |r| <= 0.5*ln(2). Given x,
-#    find r and integer k such that
-#       x = k*ln(2) + r,  |r| <= 0.5*ln(2).
-#    Here r is represented as r = hi - lo for better accuracy.
-#
-# 2. Approximate exp(r) by a special rational function on [0, 0.5*ln(2)]:
-#       R(r^2) = r*(exp(r)+1)/(exp(r)-1) = 2 + r*r/6 - r^4/360 + ...
-#
-#    A special Remez algorithm on [0, 0.5*ln(2)] is used to generate a
-#    polynomial to approximate R.
-#
-#    The computation of exp(r) thus becomes
-#                       2*r
-#       exp(r) = 1 + ----------
-#                     R(r) - r
-#                          r*c(r)
-#              = 1 + r + ----------- (for better accuracy)
-#                         2 - c(r)
-#    where
-#       c(r) = r - (P1*r^2  + P2*r^4  + ... + P5*r^10 + ...).
-#
-# 3. Scale back: exp(x) = 2^k * exp(r)
+# magic rounding constant: 1.5*2^52 Adding, then subtracting it from a float rounds it to an Int.
+# this works because eps(MAGIC_ROUND_CONST(T)) = one(T), so adding it to a smaller number alligns the lsb to the 1s place
+# values for which this trick doesn't work are goign to have outputs of 0 or Inf
+MAGIC_ROUND_CONST(::Type{Float64}) = 6.755399441055744e15
+MAGIC_ROUND_CONST(::Type{Float32}) = 1.048576f7
+
+# max, min, and subnormal arguments
+# max_exp = T(exponent_bias(T)*log(base, big(2)) + log(base, 2 - big(2.0)^-significand_bits(T)))
+MAX_EXP(n::Val{2}, ::Type{Float32}) = 128.0f0
+MAX_EXP(n::Val{2}, ::Type{Float64}) = 1024.0
+MAX_EXP(n::Val{:ℯ}, ::Type{Float32}) = 88.72284f0
+MAX_EXP(n::Val{:ℯ}, ::Type{Float64}) = 709.782712893384
+MAX_EXP(n::Val{10}, ::Type{Float32}) = 38.53184f0
+MAX_EXP(n::Val{10}, ::Type{Float64}) = 308.25471555991675
+
+# min_exp = T(-(exponent_bias(T)+significand_bits(T)) * log(base, big(2)))
+MIN_EXP(n::Val{2}, ::Type{Float32}) = -150.0f0
+MIN_EXP(n::Val{2}, ::Type{Float64}) = -1075.0
+MIN_EXP(n::Val{:ℯ}, ::Type{Float32}) = -103.97208f0
+MIN_EXP(n::Val{:ℯ}, ::Type{Float64}) = -745.1332191019412
+MIN_EXP(n::Val{10}, ::Type{Float32}) = -45.1545f0
+MIN_EXP(n::Val{10}, ::Type{Float64}) = -323.60724533877976
+
+# subnorm_exp = abs(log(base, floatmin(T)))
+# these vals are positive since it's easier to take abs(x) than -abs(x)
+SUBNORM_EXP(n::Val{2}, ::Type{Float32}) = 126.00001f0
+SUBNORM_EXP(n::Val{2}, ::Type{Float64}) = 1022.0
+SUBNORM_EXP(n::Val{:ℯ}, ::Type{Float32}) = 87.33655f0
+SUBNORM_EXP(n::Val{:ℯ}, ::Type{Float64}) = 708.3964185322641
+SUBNORM_EXP(n::Val{10}, ::Type{Float32}) = 37.92978f0
+SUBNORM_EXP(n::Val{10}, ::Type{Float64}) = 307.6526555685887
```

Comment:
Do some of these constants need to be adjusted by an ulp? The trailing 10e-5 on the `(Val{2}, Float32)` case made me curious, and checking the round-trip:
```julia
julia> TYPES = (Float32, Float64)
(Float32, Float64)

julia> FN_VALS = ((exp2, Val(2)), (exp, Val(:ℯ)), (exp10, Val(10)))
((exp2, Val{2}()), (exp, Val{:ℯ}()), (exp10, Val{10}()))

julia> import Base.Math.SUBNORM_EXP

julia> [issubnormal(fn(-SUBNORM_EXP(b, T))) for T in TYPES, (fn, b) in FN_VALS]
2×3 Matrix{Bool}:
 1  1  0
 0  0  0
```
So only the base 2 and base e Float32 constants actually exponentiate to a subnormal number. Given the `<=` in the function definitions below, should those two constants be lowered by 1-ulp (e.g. raised by 1 for the negative constants) and then the comparison changed to a `<`?

Or maybe this doesn't actually matter and/or I'm misunderstanding a detail??

Reply:
Hmm, not sure how this happened. I generated these constants programmatically and never looked at them. It looks like they should be lowered though, so I will.

Reply:
I think it's due to differences in accuracy.
```julia
julia> Float32(log(2.0, Float64(floatmin(Float32))))
-126.0f0

julia> log(2f0, floatmin(Float32))
-126.00001f0
```
It obviously gets you really close, so starting from the constant value and just iterating by `prevfloat` steps until the condition is satisifed would work as a brute-force solution.

Reply:
The other part is that the exact value of these constants doesn't matter. If it's too high, it just means that that number will take a slightly slower path.